src/nvim/mark_extended.h

```diff
@@ -0,0 +1,231 @@
+    kb_itr_next(extlines, &buf->b_extlines, &itr);\
+  }\
+  ExtMarkLine *extline;\
+  for (; kb_itr_valid(&itr); kb_itr_next(extlines, &buf->b_extlines, &itr)) { \
+    extline = kb_itr_key(&itr);\
+    if (extline->lnum > u_lnum && u_lnum !=-1) { \
+      break;\
+    }\
+      code;\
+    }
+
+// see FOR_ALL_ for documentation
+#define FOR_ALL_EXTMARKLINES_PREV(buf, l_lnum, u_lnum, code)\
+  kbitr_t(extlines) itr;\
+  ExtMarkLine t;\
+  t.lnum = u_lnum;\
+  if (!kb_itr_get(extlines, &buf->b_extlines, &t, &itr)) { \
+    kb_itr_prev(extlines, &buf->b_extlines, &itr);\
+  }\
+  ExtMarkLine *extline;\
+  for (; kb_itr_valid(&itr); kb_itr_prev(extlines, &buf->b_extlines, &itr)) { \
+    extline = kb_itr_key(&itr);\
+    if (extline->lnum < l_lnum && l_lnum !=-1) { \
+      break;\
+    }\
+    code;\
+  }
+
+// see FOR_ALL_ for documentation
+#define FOR_ALL_EXTMARKS(buf, ns, l_lnum, l_col, u_lnum, u_col, code)\
+  kbitr_t(markitems) mitr;\
+  ExtendedMark mt;\
+  mt.ns_id = ns;\
+  mt.mark_id = 0;\
+  mt.line = NULL;\
+  mt.col = l_col;\
+  FOR_ALL_EXTMARKLINES(buf, l_lnum, u_lnum, { \
+    if (!kb_itr_get(markitems, &extline->items, mt, &mitr)) { \
+        kb_itr_next(markitems, &extline->items, &mitr);\
+    } \
+    ExtendedMark *extmark;\
+    for (; kb_itr_valid(&mitr); kb_itr_next(markitems, &extline->items, &mitr)) { \
+      extmark = &kb_itr_key(&mitr);\
+      if (extmark->col > u_col && u_col !=-1) { \
+        break;\
+      }\
+      code;\
+    }\
+  })
+
+
+// see FOR_ALL_ for documentation
+#define FOR_ALL_EXTMARKS_PREV(buf, ns, l_lnum, l_col, u_lnum, u_col, code)\
+  kbitr_t(markitems) mitr;\
+  ExtendedMark mt;\
+  mt.mark_id = sizeof(uint64_t);\
+  mt.ns_id = ns;\
+  mt.col = u_col;\
+  FOR_ALL_EXTMARKLINES_PREV(buf, l_lnum, u_lnum, { \
+    if (!kb_itr_get(markitems, &extline->items, mt, &mitr)) { \
+        kb_itr_prev(markitems, &extline->items, &mitr);\
+    } \
+    ExtendedMark *extmark;\
+    for (; kb_itr_valid(&mitr); kb_itr_prev(markitems, &extline->items, &mitr)) { \
+      extmark = &kb_itr_key(&mitr);\
+      if (extmark->col < l_col && l_col !=-1) { \
+          break;\
+      }\
+      code;\
+    }\
+  })
+
+
+#define FOR_ALL_EXTMARKS_IN_LINE(items, code)\
+  kbitr_t(markitems) mitr;\
+  ExtendedMark mt;\
+  mt.ns_id = 0;\
+  mt.mark_id = 0;\
+  mt.line = NULL;\
+  mt.col = 0;\
+  if (!kb_itr_get(markitems, items, mt, &mitr)) { \
+    kb_itr_next(markitems, items, &mitr);\
+  } \
+  ExtendedMark *extmark;\
+  for (; kb_itr_valid(&mitr); kb_itr_next(markitems, items, &mitr)) { \
+    extmark = &kb_itr_key(&mitr);\
+    code;\
+  }\
+
+
+int mark_cmp(ExtendedMark a, ExtendedMark b);
+
+#define markitems_cmp(a, b) (mark_cmp((a), (b)))
+KBTREE_INIT(markitems, ExtendedMark, markitems_cmp, 10)
+
+
+typedef struct ExtMarkLine
+{
+  linenr_T lnum;
+  kbtree_t(markitems) items;
+} ExtMarkLine;
+
+
+KBTREE_INIT(extlines, ExtMarkLine *, extline_cmp, 10)
+
+
+static uint64_t EXTMARK_NAMESPACE_COUNTER = 0;
```

Comment:
ah was unable to get this working, Cold you show me how? I gave permissions to commit to the pr